StorageUtility/Linux/StorageUtility.cpp

```diff
@@ -29,239 +32,172 @@ namespace vtStor
+    {
+        const char *devicePath;
 
-        PSP_INTERFACE_DEVICE_DETAIL_DATA pDevDetailData = (PSP_INTERFACE_DEVICE_DETAIL_DATA)HeapAlloc( GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, bufferSize );
-        if ( NULL == pDevDetailData )
-        {
-            if ( eOnErrorBehavior::Continue == OnErrorBehavior ) { continue; }
-            SetupDiDestroyDeviceInfoList( devInfoHandle );
-            return( eErrorCode::Memory );
-        }
-        pDevDetailData->cbSize = sizeof( SP_INTERFACE_DEVICE_DETAIL_DATA );
+        devicePath = udev_list_entry_get_name(udevListEntry);
+        udevDevice = udev_device_new_from_syspath(udevObject, devicePath);
+        String udeviceNode = udev_device_get_devnode(udevDevice);
 
-        if ( SetupDiGetDeviceInterfaceDetail( devInfoHandle, &devInterfaceData, pDevDetailData, bufferSize, NULL, NULL ) == FALSE )
+        if (true == IsStorageDisk(udevDevice))
         {
-            error = GetLastError();
-            if ( ERROR_INSUFFICIENT_BUFFER != error )
-            {
-                HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
-                if ( eOnErrorBehavior::Continue == OnErrorBehavior ) { continue; }
-                SetupDiDestroyDeviceInfoList( devInfoHandle );
-                return( eErrorCode::Unknown );
-            }
+            DevicePaths.push_back(udeviceNode);
+            SysDevicePaths.push_back(devicePath);
         }
 
-        eErrorCode errorCode;
-        errorCode = eErrorCode::Unknown;
-        Callback.Function(Callback.Data, devInfoHandle, devInfoData, devInterfaceData, pDevDetailData, bufferSize, errorCode);
-
-        HeapFree(GetProcessHeap(), HEAP_NO_SERIALIZE, pDevDetailData);
+        udev_device_unref(udevDevice);
     }
+    //! Free the enumerator object
+    udev_enumerate_unref(enumerate);
 
-    SetupDiDestroyDeviceInfoList( devInfoHandle );
+    udev_unref(udevObject);
 
-    return( eErrorCode::None );
-}*/
+    return(eErrorCode::None);
+}
 
-eErrorCode GetStorageAdapterProperty( HANDLE Handle, sStorageAdapterProperty& AdapterProperty )
+eErrorCode GetAdapterBus(DeviceHandle& Handle, String SysDevicePath)
 {
-   /* STORAGE_DESCRIPTOR_HEADER storageDescHeader;
-
-    STORAGE_PROPERTY_QUERY StorageProperty;
-    StorageProperty.PropertyId = StorageAdapterProperty;
-    StorageProperty.QueryType = PropertyStandardQuery;
-
-    DWORD bytesReturned;
-    DWORD ret =
-        DeviceIoControl(
-                        Handle,
-                        IOCTL_STORAGE_QUERY_PROPERTY,
-                        &StorageProperty,
-                        sizeof(STORAGE_PROPERTY_QUERY),
-                        &storageDescHeader,
-                        sizeof(STORAGE_DESCRIPTOR_HEADER),
-                        &bytesReturned,
-                        NULL
-                        );
-    if ( 0 == ret )
+    struct udev *udevObject;
+    struct udev_device *udevDevice;
+
+    udevObject = udev_new();
+    if (nullptr == udevObject)
```

Comment:
if (nullptr == udevObject) then return to unknown, is udev_new allocate a memory?
Why don't return eErrorCode::Memory?